### PR TITLE
Support invoice expiry over a year

### LINKED ISSUE #1235: Drop one-year limit on invoice expiry
We shouldn't fail to parse just because an invoice is kinda dumb. This should be more practical now that we switched to `Duration` instead of `SystemTime`.

### PR DESCRIPTION
**Pull request by jkczyz**
The lightning-invoice crate represents timestamps as `Duration` since the UNIX epoch rather than a `SystemTime`. Therefore, internal calculations are in terms of `u64`-based `Duration`s. This allows for relaxing the one year maximum expiry.

Fixes #1235

### CODE REVIEW

File: lightning-invoice/src/lib.rs [hunk 1]
```diff
@@ -85,14 +85,16 @@ mod sync;
...
-/// one of the unit tests, please run them.
-const SYSTEM_TIME_MAX_UNIX_TIMESTAMP: u64 = core::i32::MAX as u64;
+/// The maximum timestamp that can be represented as a [`Duration`] since the UNIX epoch while
+/// allowing for adding an expiry without overflowing.
+const MAX_TIMESTAMP: u64 = core::i64::MAX as u64;
 
-/// Allow the expiry time to be up to one year. Since this reduces the range of possible timestamps
-/// it should be rather low as long as we still have to support 32bit time representations
-const MAX_EXPIRY_TIME: u64 = 60 * 60 * 24 * 356;
+/// The maximum expiry allowed, represented as a [`Duration`] since the invoice timestamp.
```
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:92]: Why have a max expiry at all? Can we just update the expiry-check logic to do a saturating add?

Comment (jkczyz) [reply lightning-invoice/src/lib.rs]: Good idea!

**Comment by TheBlueMatt** [comment]
I assume the serialization of `PositiveTimestamp` needs updating?

```
impl ToBase32 for PositiveTimestamp {
	fn write_base32<W: WriteBase32>(&self, writer: &mut W) -> Result<(), <W as WriteBase32>::Err> {
		// FIXME: use writer for int encoding
		writer.write(
			&try_stretch(encode_int_be_base32(self.as_unix_timestamp()), 7)
				.expect("Can't be longer due than 7 u5s due to timestamp bounds")
		)
	}
}
```

File: lightning-invoice/src/lib.rs [hunk 2]
```diff
@@ -85,14 +85,12 @@ mod sync;
...
-const SYSTEM_TIME_MAX_UNIX_TIMESTAMP: u64 = core::i32::MAX as u64;
+/// The number of bits used to represent timestamps as defined in BOLT 11.
+const TIMESTAMP_BITS: usize = 35;
 
-/// Allow the expiry time to be up to one year. Since this reduces the range of possible timestamps
-/// it should be rather low as long as we still have to support 32bit time representations
-const MAX_EXPIRY_TIME: u64 = 60 * 60 * 24 * 356;
+/// The maximum timestamp that can be represented as a [`Duration`] since the UNIX epoch while
+/// allowing for adding an expiry without overflowing.
+const MAX_TIMESTAMP: u64 = core::u64::MAX >> (64 - TIMESTAMP_BITS);
```
Comment (devrandom) [inline lightning-invoice/src/lib.rs:93]: a bit confusing.  isn't `(1 << TIMESTAMP_BITS) - 1` more idiomatic?

also, comment says "allowing for adding an expiry" - what is the max expiry time and should we subtract it from the above?

File: lightning-invoice/src/de.rs [hunk 3]
```diff
@@ -359,7 +359,6 @@ impl FromBase32 for PositiveTimestamp {
 			.expect("7*5bit < 64bit, no overflow possible");
 		match PositiveTimestamp::from_unix_timestamp(timestamp) {
 			Ok(t) => Ok(t),
-			Err(CreationError::TimestampOutOfBounds) => Err(ParseError::TimestampOverflow),
```
Comment (devrandom) [inline lightning-invoice/src/de.rs:362]: why delete this branch - this error could still be returned, no?

File: lightning-invoice/src/lib.rs (hunk 2 above)
Comment (jkczyz) [reply lightning-invoice/src/lib.rs:93]: > a bit confusing. isn't `(1 << TIMESTAMP_BITS) - 1` more idiomatic?
> 

Yes, thanks! I temporarily used that in another PR in the context of log approximation, but had blanked on the idiom after dropping the commit. 😛 

> also, comment says "allowing for adding an expiry" - what is the max expiry time and should we subtract it from the above?

We dropped the maximum expiry, simply allowing anything that fits into a `u64` and using a saturating add if we ever overflow. The `u64` space is enormous, so it should never be a problem in reality. Updated the comment.

**Comment by jkczyz** [comment]
> I assume the serialization of `PositiveTimestamp` needs updating?
> 
> ```
> impl ToBase32 for PositiveTimestamp {
> 	fn write_base32<W: WriteBase32>(&self, writer: &mut W) -> Result<(), <W as WriteBase32>::Err> {
> 		// FIXME: use writer for int encoding
> 		writer.write(
> 			&try_stretch(encode_int_be_base32(self.as_unix_timestamp()), 7)
> 				.expect("Can't be longer due than 7 u5s due to timestamp bounds")
> 		)
> 	}
> }
> ```

Discussed offline. BOLT 11 defines timestamps as 35 bits, so updated the PR yesterday to enforce this instead.

File: lightning-invoice/src/lib.rs [hunk 4]
```diff
@@ -556,10 +487,7 @@ impl<D: tb::Bool, H: tb::Bool, T: tb::Bool, C: tb::Bool, S: tb::Bool> InvoiceBui
 
 	/// Sets the expiry time
 	pub fn expiry_time(mut self, expiry_time: Duration) -> Self {
-        match ExpiryTime::from_duration(expiry_time) {
-            Ok(t) => self.tagged_fields.push(TaggedField::ExpiryTime(t)),
-            Err(e) => self.error = Some(e),
-        };
+        self.tagged_fields.push(TaggedField::ExpiryTime(ExpiryTime::from_duration(expiry_time)));
```
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:490]: nit: when you squash, can you make this a tab not space (it was already spaces before).

**Review by TheBlueMatt — APPROVED** [review]
ACK 3b14a76d0114f6f8

**Comment by devrandom** [comment]
ACK

**Review by devrandom — APPROVED** [review]

**Review by valentinewallace — APPROVED** [review]

**Pull request merged**